lib/actions/reassemble.js

```diff
@@ -54,19 +87,196 @@ async function processAction(msg) {
     Currently the group has ${messagesNumberSeen} of ${messageGroupSize} message(s).`,
   );
 
-  if (messagesNumberSeen >= messageGroupSize) {
-    parsedMessageGroup.messages.forEach((message) => {
-      incomingData[message.messageId] = message.messageData;
-    });
+  // when grouSized option is selected
+  if (mode === 'groupSize') {
+    if (messagesNumberSeen >= messageGroupSize) {
+      parsedMessageGroup.messages.forEach((message) => {
+        incomingData[message.messageId] = message.messageData;
+      });
+      await this.emit('data', messages.newMessageWithBody({
+        groupSize,
+        groupId,
+        messageData: incomingData,
+      }));
+      await storage.deleteObjectById(messageGroupId);
+      this.logger.info(`Message group with id ${messageGroupId} has been deleted`);
+    }
+  }
 
-    await this.emit('data', messages.newMessageWithBody({
-      groupSize,
-      groupId,
-      messageData: incomingData,
-    }));
-    await storage.deleteObjectById(messageGroupId);
-    this.logger.info(`Message group with id ${messageGroupId} has been deleted`);
+  // When delay timer option is selected
+  if (mode === 'timeout') {
+    delay = msg.body.timersec;
+    delay = (msg.body.timersec >= 40000) ? 40000 : msg.body.timersec;
+    clearTimeout(timeHandle);
+    timeHandle = setTimeout(timer, delay, this);
+    if (groupList.includes(messageGroupId)) {
+      // no action required
+    } else {
+      groupList.push(messageGroupId);
+    }
+  }
+
+  // When both groupSize and delay timer option is selected
+  if (mode === 'groupSize&timeout') {
+    delay = msg.body.timersec;
+    delay = (msg.body.timersec >= 40000) ? 40000 : msg.body.timersec;
+    clearTimeout(timeHandle);
+    timeHandle = setTimeout(timer, delay, this);
+
+    if (groupList.includes(messageGroupId)) {
+      // no action required
+    } else {
+      groupList.push(messageGroupId);
+    }
+
+    if (messagesNumberSeen >= messageGroupSize) {
+      parsedMessageGroup.messages.forEach((message) => {
+        incomingData[message.messageId] = message.messageData;
+      });
+
+      await this.emit('data', messages.newMessageWithBody({
+        groupSize,
+        groupId,
+        messageData: incomingData,
+      }));
+      await storage.deleteObjectById(messageGroupId);
+      this.logger.info(`Message group with id ${messageGroupId} has been deleted`);
+      groupList = groupList.filter((def) => def !== messageGroupId);
+    }
```

Comment:
If the `mode === 'groupSize&timeout'` code branch is equivalent to the `mode === 'timeout'` code branch followed by the `mode === 'timeout'` code branch than why do you need to maintain a 3rd code branch?

Reply:
1st code branch - Produce Groups of Fixed Size (Don't Emit Partial Groups): equivalent to `'groupSize'`
2nd code branch -  Group All Incoming Messages: equivalent to `'timeout'`
3rd code branch - Produce Groups of Fixed Size (Emit Partial Groups): equivalent to `'timeout'` followed by `'groupSize'`

This 3 variants were the requirements given for this branch. 
SHould we change that?

Reply:
I would be in favor of reducing copy & pasted code but this might just be my opinion.

Reply:
After a discussion, we decided to leave the three paths as 3 explicit paths.